packages/docusaurus-theme-classic/src/theme/hooks/usePlatformValue.ts

```diff
@@ -0,0 +1,34 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import useWindowSize from '@theme/hooks/useWindowSize';
+import type {PlatformDependentConfig} from '@docusaurus/types';
+
+// TODO I should be able to use generics here...
+function isSimpleValue(
+  option: string | number | boolean | Record<string, string | number | boolean>,
+): option is string | number | boolean {
+  return typeof option !== 'object';
+}
+
+function usePlatformValue<T extends string | number | boolean>(
+  option: PlatformDependentConfig<T> | undefined,
+): T | undefined {
+  const windowSize = useWindowSize();
```

Comment:
This is not possible to do this without introducing hydration layout shifts.

What is rendered on the server must stay the same once React hydrates on the client.

A UI element with `mobile: true` & `desktop: false` should not lead to the title magically appearing/disappearing after JS loads. 

You can't rely on the `windowSize` dynamic value, only on the config/frontmatter. This wouldn't work on the server. Remember also that Docusaurus may want to work better without React hydration in the future, so don't assume any JS will run on the client-side: it should work with JS disabled.

To avoid hydration layout shifts, we must play with CSS mediaqueries and also likely render twice the same element in the DOM with different mediaqueries.

This is not a simple problem to solve on a static website, but there are some solutions in the SSR land (https://artsy.github.io/blog/2019/05/24/server-rendering-responsively/)

For these reasons I'm not really in favor of supporting this mobile/desktop config for all the possible frontmatters, but only those for which there are a strong use-case.

Reply:
There's a pretty strong use-case for most `hide_*` front matter. You mentioned the TOC one somewhere, and we agreed on the `hide_sidebar`. I think SSR can be handled gracefully here. Maybe the idea of `usePlatformValue` is still too generic

Reply:
We can definitively implement this but this `usePlatformValue` is a trap and using it will create layout shifts, we it must be done in another way, without this abstraction, and consider SSR properly everywhere such config is used.

> and we agreed on the hide_sidebar

Yes, and the sidebar is already rendered differently for mobile/desktop, the complexity is already there so adding this feature in such a case is less complex than introducing it for elements that render the same on mobile/desktop

Reply:
I think the Slorber criticism is mostly of a fact that platform should not change during the visit on the site, for instance if you switch to "desktop mode" on mobile, the entire site should be refetched. And as such it doesn't make sense for the usePlatformValue to be a hook. But than we are going back to my original question, how should we determinate the platform.

Btw @slorber I think the media query are also based of off window size so you might experience layout shifts with them as well. Is "hydration layout shift" some specific problem?

Reply:
There's no way to know which platform it is before arriving at the browser - you can actually determinate which platform it is before arriving at the browser, it is just not window size based. With high DPI phones the size based approach is a bit flawed anyway IMHO. Most modern browsers will send the platform indicating string to the server when fetching the site.

Reply:
> There's no way to know which platform it is before arriving at the browser

True, but the critical CSS is inlined into the static HTML pages, so the browser has all the CSS before painting your HTML page, and mediaqueries allow to hide/show elements even before the React code has finished loading, as soon as the first pixels are displayed to the user screen.

> The blog link Sebastien always sends: joshwcomeau.com/react/the-perils-of-rehydration

This post is about a different problem due to server/browser JSX mismatches, not really related to layout shifts.

---

TLDR: the feature you implement should rather work in a production build with JS disabled (there are chrome plugins to disable JS, and it is time-consuming to test as you really need to build the prod site).

A good source of inspiration is the `@theme/ThemedImage` component: you can see the complexity introduced to handle layout shifts:

```jsx
import React from 'react';
import clsx from 'clsx';

import useIsBrowser from '@docusaurus/useIsBrowser';
import useThemeContext from '@theme/hooks/useThemeContext';
import type {Props} from '@theme/ThemedImage';

import styles from './styles.module.css';

const ThemedImage = (props: Props): JSX.Element => {
  const isBrowser = useIsBrowser();
  const {isDarkTheme} = useThemeContext();
  const {sources, className, alt = '', ...propsRest} = props;

  type SourceName = keyof Props['sources'];

  const clientThemes: SourceName[] = isDarkTheme ? ['dark'] : ['light'];

  const renderedSourceNames: SourceName[] = isBrowser
    ? clientThemes
    : // We need to render both images on the server to avoid flash
      // See https://github.com/facebook/docusaurus/pull/3730
      ['light', 'dark'];

  return (
    <>
      {renderedSourceNames.map((sourceName) => (
        <img
          key={sourceName}
          src={sources[sourceName]}
          alt={alt}
          className={clsx(
            styles.themedImage,
            styles[`themedImage--${sourceName}`],
            className,
          )}
          {...propsRest}
        />
      ))}
    </>
  );
};

export default ThemedImage;
```

```
.themedImage {
  display: none;
}

html[data-theme='light'] .themedImage--light {
  display: initial;
}

html[data-theme='dark'] .themedImage--dark {
  display: initial;
}
```

You can see that we need to render in both light/dark version on the server, so that it does not show first the light version and then updates to the dark version.

(the `html[data-theme='light']` acts a bit like a mediaquery here because we set that data-attribute as soon as possible with js inlined in the static HTML files to avoid theme flashes => a problem that is quite similar to layout shifts in practice)